### PR TITLE
did typescript semi conversion,  I tried to make simple screen casts about the experience

### PR DESCRIPTION
**Pull request by SteveStrong**
how I started

http://screencast.com/t/EPbHbZ8f

getting things to compile

http://screencast.com/t/Vm8X2KPj8

tricks on using external libs.

http://screencast.com/t/DlSlyZzYATdG

### CODE REVIEW

File: AllReadyApp/Web-App/AllReady/Areas/Admin/Views/Activity/Edit.cshtml [hunk 1]
```diff
@@ -154,7 +154,7 @@
     <script type="text/javascript">
 
             System.config({
-                baseURL: location.protocol + "//" + location.host + '/ts/'
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/Areas/Admin/Views/Activity/Edit.cshtml]: I would tend to push this stuff down a level. So for instance on my projects I tend to have either a folder per controller with the associated files for views off that controller
image

or I'll inset another level so that I can have things like Controls, Pages, Infrastructure in the scripts directory

image

Obviously each file is namespaced in accordance with the hierarchy. Arranging javascript should be just like arranging "real" c# code.

File: AllReadyApp/Web-App/AllReady/Views/Campaign/Map.cshtml [hunk 2]
```diff
@@ -22,7 +22,7 @@
     var modelCampaign = @Json.Serialize(Model);
 
     System.config({
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/Views/Campaign/Map.cshtml]: We seem to be duplicating this system configuration in a lot of files. Perhaps we could centralize it into one file. We could load it in the _Layout

File: AllReadyApp/Web-App/AllReady/Views/Shared/_LayoutMainPage.cshtml [hunk 3]
```diff
@@ -2,6 +2,7 @@
 <html>
 <head>
     <meta charset="utf-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/Views/Shared/_LayoutMainPage.cshtml]: This is a good call but maybe should be part of a different PR.

File: AllReadyApp/Web-App/AllReady/gulpfile.js [hunk 4]
```diff
@@ -49,13 +58,81 @@ gulp.task("watch", function () {
...
+
+    var allTypeScript = 'ts/**/*.ts';
+    var libraryTypeScriptDefinitions = './tools/typings/**/*.ts';
+    var sourceTsFiles = [allTypeScript,                //path to typescript files
+                         libraryTypeScriptDefinitions]; //reference to library .d.ts files
+
+
+    return gulp.src(sourceTsFiles)
+        .pipe(tsc({
+            noImplicitAny: false,
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/gulpfile.js]: We should work towards turning this to true.

File: AllReadyApp/Web-App/AllReady/gulpfile.js [hunk 5]
```diff
@@ -49,13 +58,81 @@ gulp.task("watch", function () {
...
-//http://weblogs.asp.net/dwahlin/creating-a-typescript-workflow-with-gulp
-var tsProject = ts.createProject('tsconfig.json');
 
+//https://www.npmjs.com/package/gulp-typescript
 gulp.task('typescript:Build', function () {
-    var tsResult = tsProject.src() // instead of gulp.src(...) 
-        .pipe(ts(tsProject));
-    return tsResult.js.pipe(gulp.dest('wwwroot/release'));
+
+    var allTypeScript = 'ts/**/*.ts';
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/gulpfile.js]: To me all these definitions should be sucked up and put into a config object. 

`
paths.typescript = {
    source: 'ts/**/*.ts',
    definitions: 'tools/typings/**/*.ts'
}
`

File: AllReadyApp/Web-App/AllReady/gulpfile.js [hunk 6]
```diff
@@ -49,13 +58,81 @@ gulp.task("watch", function () {
...
+
 });
 
+
+
+//https://www.npmjs.com/package/gulp-typescript
+//http://weblogs.asp.net/dwahlin/creating-a-typescript-workflow-with-gulp
+var tsProject = tsc.createProject('./tsconfig.json');
+
+//gulp.task('typescript:Build', function () {
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/gulpfile.js]: These all look like good targets, I'm not sure why they're commented out

File: AllReadyApp/Web-App/AllReady/package.json [hunk 7]
```diff
@@ -9,7 +9,9 @@
     "gulp-uglify": "1.2.0",
     "bower": "1.6.5",
     "rimraf": "2.2.8",
+    "gulp-tslint": "3.6.0",
     "gulp-typescript": "2.9.2",
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/package.json]: We should take a dependency on typescript and pass it into gulp-typescript so that we don't rely on people having typescript 1.6.X installed on their machines.

File: AllReadyApp/Web-App/AllReady/wwwroot/js/activityDetails.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
-﻿var HTBox;
-(function (HTBox) {
+var HTBox;
+(function (HTBox, $) {
```
Comment (stimms) [inline AllReadyApp/Web-App/AllReady/wwwroot/js/activityDetails.js]: I don't like passing in a variable called `$` on jQuery projects that has a pretty specific meaning. Perhaps `container` instead.

**Comment by stimms** [comment]
.js files and .map.js files should not be checked in. They are build artifacts and should be checked in no more than .dll and .pdb files.

**Comment by mheggeseth** [comment]
There should probably be a `ko.filterableList.ts` file but I don't see it in the commit list.

**Comment by dpaquette** [comment]
One thing to do before this can be merged to ensure the typescript files are compiled as part of the prepublish command in project.json.

```
"scripts": {
    "prepublish": [ "npm install", "bower install", "gulp clean", "gulp typescript:Build", "gulp min" ]
  }
```

**Comment by dpaquette** [comment]
We should also have the gulp watch task setup to run the typescript:Build task anytime a typescript file changes.This would ensure that the wwwroot/js file is always up to date for developers.

Since the files in wwwroot/js are always auto generated we should eventually be removing those from source control. We don't have to do that for this PR but something to keep in mind.

**Comment by stimms** [comment]
Okay can we merge this and then I'll submit follow up PRs? Other that @dpaquette's concerns I think everything else can be deferred.

**Comment by BillWagner** [comment]
Will do!
